### ISSUE TITLE
[java] False positive with guava preconditions

### ISSUE THREAD

**Issue by hgschmie** [pmd/pmd#5209]
**Affects PMD Version:**  7.5.0 (only version tested). Did work in 6.55.0

**Rule:** CheckResultSet

Please provide the rule name and a link to the rule documentation:
https://pmd.github.io/pmd/pmd_rules_java_bestpractices.html#checkresultset

**Description:**

Calling `ResultSet#next()` in a Guava precondition reports `Rule:CheckResultSet Priority:3 Always check the return of one of the navigation method (next,previous,first,last) of a ResultSet..`

**Code Sample demonstrating the issue:**

https://github.com/hgschmie/pg-embedded/blob/master/src/main/java/de/softwareforge/testing/postgres/embedded/EmbeddedPostgres.java#L495

```java
try (Connection c = createDefaultDataSource().getConnection();
        Statement s = c.createStatement();
        ResultSet rs = s.executeQuery("SELECT 1")) {
    checkState(rs.next(), "expecting single row");
    checkState(rs.getInt(1) == 1, "expecting 1 as result");
    checkState(!rs.next(), "expecting single row");
    return true;
}
```

**Expected outcome:**

code is correct.

```



**Running PMD through:** *[CLI | Ant | Maven | Gradle | Designer | Other]*

[ This is a possible duplicate of #5031 ]

**Comment by jsotuyod** [comment]
@hgschmie no, this is not a duplicate of that issue. This is something else entirely.

The rule expects the call to be within a return statement / if / while. #5031 fails because do-while is not currently considered.

This case on the other hand fails because detecting this would require PMD to know the semantics of Guava's `Preconditions.checkState`.

**Comment by hgschmie** [comment]
Thanks for looking into this so quickly. So is this basically a "works as expected" because it does not adhere to the pattern that pmd understands.

And why did it not flag a problem in 6.55 ?

**Comment by jsotuyod** [comment]
The short answer is, because PMD 7 is far more powerful and capable than PMD 6 ever was…

Prior to PMD 7, PMD's Java grammar was very cumbersome and somewhat distant to the Java Language Specification, and our symbol table and type resolution capabilities were limited. These factor combined into making a lot of rules really struggle to properly capture all scenarios. In this particular rule's case, the rule would not consider `ResultSet` instances that were declared in a try-with-resources (as is your case). So it's not that PMD 6.55 properly handled this scenario, it was on the contrary, completely ignoring it, and would not have flagged even if you had called `rs.next()` outside of `checkState`.

Maybe it's worth extending the rule to allow ignoring calls being passed to methods matching a given signature…this would support Guava, Apache Commons Lang's `Validate`, and any other there may be…
